crossvalidation.py

```diff
@@ -2,123 +2,188 @@
+        """,
+    )
+
+    parser.add_argument(
+        "-H",
+        "--use_hyperopt",
+        action="store_true",
+        help="""Switch to tune parameters using hyperopt instead of grid search.
+    """,
+    )
+
+    parser.add_argument(
+        "-e",
+        "--max_evals",
+        type=int,
+        default=100,
+        help="""Maximum number of evaluations for hyperopt.""",
+    )
+
+    parser.add_argument(
+        "-i",
+        "--max_iter",
+        type=int,
+        default=100,
+        help="""Maximum number of iterations for model optimization.""",
     )
-    test_loss = pf.LossFunctional(test_sd, control, params['regularization'])
+
+    return parser.parse_args()
+
+
+def objective(tuning_params = None):
+    # defaults
+    params = {
+        "method": "BFGS",
+        "options": {
+            'gtol': 1e-8,
+            'xrtol': 1e-8,
+            'maxiter': args.max_iter,
+        },
+        "boundary": None,
+    }
+
+    paramsfile = args.json
+    outdir = os.path.dirname(paramsfile)
+    with open(paramsfile, 'r') as f:
+        file_params = json.load(f)
+
+    params.update(file_params)
     
-    invp = pf.InverseProblem(eqn, train_loss)
-    m_hats, losses, optim_return = invp.optimize(
-            control, 
-            method=params['method'],
-            options=params['options'],
+    if args.use_hyperopt:
+        results_file = os.path.join(outdir, 'results_hyperopt.pkl')
+    else:
+        results_file = os.path.join(outdir, 'results.pkl')
+
+    # load spatial and genetic data
+    spatial_data = pd.read_csv(os.path.join(outdir, params['spatial_data'])).rename(
+            columns={"site_name": "name", "long": "x", "lat": "y"}
+    )
+    genetic_data = pd.read_csv(os.path.join(outdir, params['genetic_data'])).rename(
+            columns={"loc1": "name1", "loc2": "name2", "dxy": "divergence"}
     )
-    results[fold]['m_hats'] = m_hats
-    results[fold]['losses'] = losses
-    results[fold]['optim_return'] = optim_return
+    data = inference.SpatialDivergenceData(spatial_data, genetic_data)
+    data.normalise(min_xy=0.2, max_xy=0.8)
+
+    # optionally determine the boundary parameters
+    if params['boundary'] is None:
+        bdry_params = data.choose_epsilons()
+        params['boundary'] = {
+                "epsilon": bdry_params['eps0'],
+                "eps0": bdry_params['eps0'],
+                "eps1": bdry_params['eps1']
+        }
+
+    boundary = data.boundary_fn(eps0=params['boundary']['eps0'], eps1=params['boundary']['eps1'])
+
+    if len(params['mesh']) == 1 and isinstance(params['mesh'], str):
+        mesh = fenics.Mesh(params['mesh'])
+    elif len(params['mesh']) == 2:
+        mesh = fenics.UnitSquareMesh(*params['mesh'])
+    else:
+        print(f"mesh must be an xml file name or (width, height), got {params['mesh']}")
+        sys.exit()
+    V = fenics.FunctionSpace(mesh, 'CG', 1)
+    W = pf.ProductFunctionSpace(V)
+
```

Comment:
in fact, everything until here is set-up that ought to beoutside of the objective function